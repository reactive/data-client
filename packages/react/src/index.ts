Object.hasOwn =
  Object.hasOwn ||
  /* istanbul ignore next */ function hasOwn(it, key) {
    return Object.prototype.hasOwnProperty.call(it, key);
  };
export { Controller, ExpiryStatus, actionTypes } from '@data-client/core';
export type {
  EndpointExtraOptions,
  FetchFunction,
  ResolveType,
  EndpointInterface,
  EntityInterface,
  Queryable,
  SchemaArgs,
  Schema,
  SchemaClass,
  DenormalizeNullable,
  Denormalize,
  Normalize,
  NormalizeNullable,
  FetchAction,
  InvalidateAction,
  UnsubscribeAction,
  SubscribeAction,
  ResetAction,
  SetAction,
  SetResponseAction,
  ActionTypes,
  NetworkError,
  UnknownError,
  ErrorTypes,
  AbstractInstanceType,
  UpdateFunction,
  State,
  PK,
  Dispatch,
  Middleware,
  MiddlewareAPI,
  Manager,
  GCInterface,
  GCOptions,
  CreateCountRef,
  // used in Controller generic
  DataClientDispatch,
  GenericDispatch,
} from '@data-client/core';
export { default as GCPolicy } from './state/GCPolicy.js';
export * from './managers/index.js';
export * from './components/index.js';
export * from './hooks/index.js';
export { StateContext, ControllerContext, StoreContext } from './context.js';
export type { Store } from './context.js';
export * as __INTERNAL__ from './internal.js';
export { usePromisifiedDispatch } from '@data-client/use-enhanced-reducer';
