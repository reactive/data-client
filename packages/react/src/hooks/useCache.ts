import type {
  EndpointInterface,
  DenormalizeNullable,
  Schema,
  FetchFunction,
  ResolveType,
  NI,
} from '@data-client/core';
import { ExpiryStatus } from '@data-client/core';
import { useEffect, useMemo } from 'react';

import useCacheState from './useCacheState.js';
import useController from '../hooks/useController.js';

/**
 * Read an Endpoint's response if it is ready.
 *
 * `useCache` guarantees referential equality globally.
 * @see https://dataclient.io/docs/api/useCache
 */
export default function useCache<
  E extends Pick<
    EndpointInterface<FetchFunction, Schema | undefined, undefined | boolean>,
    'key' | 'schema' | 'invalidIfStale'
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E['key']>] | readonly [null]
): E['schema'] extends undefined | null ?
  E extends (...args: any) => any ?
    ResolveType<E> | undefined
  : any
: DenormalizeNullable<E['schema']> {
  const state = useCacheState();
  const controller = useController();

  const key = args[0] !== null ? endpoint.key(...args) : '';
  const cacheResults = key && state.endpoints[key];
  const meta = state.meta[key];

  // Compute denormalized value
  const { data, expiryStatus, expiresAt, countRef } = useMemo(() => {
    return controller.getResponseMeta(endpoint, ...args, state);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    cacheResults,
    state.indexes,
    state.entities,
    state.entitiesMeta,
    meta,
    key,
  ]);

  // If we are hard invalid we must fetch regardless of triggering or staleness
  const forceFetch = expiryStatus === ExpiryStatus.Invalid;

  /*********** This block is to ensure results are only filled when they would not suspend **************/
  // This computation reflects the behavior of useSuspense/useFetch
  // It only changes the value when expiry or params change.
  // This way, random unrelated re-renders don't cause the concept of expiry
  // to change
  const expired = useMemo(() => {
    return (Date.now() > expiresAt || forceFetch) && key;
    // we need to check against serialized params, since params can change frequently
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [expiresAt, key, forceFetch, state.lastReset]);

  // fully "valid" data will not suspend/loading even if it is not fresh
  const loading = expiryStatus !== ExpiryStatus.Valid && expired;
  /****************************************************************************************************/

  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(countRef, [data]);

  return useMemo(() => {
    // if useSuspense() would suspend, don't include entities from cache
    if (loading) {
      if (!endpoint.schema) return undefined;
      // TODO: use getResponse() once it just returns data
      return controller.getResponseMeta(endpoint, ...args, {
        ...state,
        entities: {},
      }).data as any;
    }
    return data;
    // key substitutes args + endpoint
    // we only need cacheResults, as entities are not used in this case
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [key, data, loading, cacheResults]);
  /*********************** end block *****************************/
}
