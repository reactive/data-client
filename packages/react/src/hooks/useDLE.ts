import type {
  Denormalize,
  DenormalizeNullable,
  ErrorTypes,
  EndpointInterface,
  FetchFunction,
  Schema,
  ResolveType,
} from '@data-client/core';
import { ExpiryStatus } from '@data-client/core';
import { useEffect, useMemo } from 'react';

import useCacheState from './useCacheState.js';
import useController from './useController.js';

type SchemaReturn<S extends Schema | undefined> =
  | {
      data: Denormalize<S>;
      loading: false;
      error: undefined;
    }
  | { data: DenormalizeNullable<S>; loading: true; error: undefined }
  | { data: DenormalizeNullable<S>; loading: false; error: ErrorTypes };

type AsyncReturn<E> =
  | {
      data: E extends (...args: any) => any ? ResolveType<E> : any;
      loading: false;
      error: undefined;
    }
  | { data: undefined; loading: true; error: undefined }
  | { data: undefined; loading: false; error: ErrorTypes };

/**
 * Use async data with { data, loading, error } (DLE)
 * @see https://dataclient.io/docs/api/useDLE
 */
export default function useDLE<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>]
): E['schema'] extends undefined | null ? AsyncReturn<E>
: SchemaReturn<E['schema']>;

export default function useDLE<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>] | readonly [null]
): {
  data: E['schema'] extends undefined | null ? undefined
  : DenormalizeNullable<E['schema']>;
  loading: boolean;
  error: ErrorTypes | undefined;
};

export default function useDLE<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(endpoint: E, ...args: readonly [...Parameters<E>] | readonly [null]): any {
  const state = useCacheState();
  const controller = useController();

  const key = args[0] !== null ? endpoint.key(...args) : '';
  const cacheResults = key && state.endpoints[key];
  const meta = state.meta[key];

  // Compute denormalized value
  // eslint-disable-next-line prefer-const
  let { data, expiryStatus, expiresAt, countRef } = useMemo(() => {
    return controller.getResponseMeta(endpoint, ...args, state);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    cacheResults,
    state.indexes,
    state.entities,
    state.entitiesMeta,
    meta,
    key,
  ]);

  // If we are hard invalid we must fetch regardless of triggering or staleness
  const forceFetch = expiryStatus === ExpiryStatus.Invalid;

  const maybePromise = useMemo(() => {
    // null params mean don't do anything
    if ((Date.now() <= expiresAt && !forceFetch) || !key) return;

    return controller.fetch(endpoint, ...(args as any)).catch(() => {});
    // we need to check against serialized params, since params can change frequently
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [expiresAt, key, forceFetch, state.lastReset]);

  // fully "valid" data will not suspend/loading even if it is not fresh
  const loading = expiryStatus !== ExpiryStatus.Valid && !!maybePromise;

  data = useMemo(() => {
    // if useSuspense() would suspend, don't include entities from cache
    if (loading) {
      if (!endpoint.schema) return undefined;
      // TODO: use getResponse() once it just returns data
      return controller.getResponseMeta(endpoint, ...args, {
        ...state,
        entities: {},
      }).data as any;
    }
    return data;
    // key substitutes args + endpoint
    // we only need cacheResults, as entities are not used in this case
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [key, data, loading, cacheResults]);

  const error = controller.getError(endpoint, ...args, state);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(countRef, [data]);

  return {
    data,
    loading,
    error,
  };
}
