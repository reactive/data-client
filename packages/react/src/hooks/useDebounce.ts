import { useEffect, useState, useTransition } from 'react';

/**
 * Keeps value updated after delay time
 *
 * @see https://dataclient.io/docs/api/useDebounce
 * @param value Any immutable value
 * @param delay Time in miliseconds to wait til updating the value
 * @param updatable Whether to update at all
 * @example
 ```
 const [debouncedQuery, isPending] = useDebounce(query, 200);
 const list = useSuspense(getThings, { query: debouncedQuery });
 ```
 */
export default function useDebounce<T>(
  value: T,
  delay: number,
  updatable = true,
): [T, boolean] {
  const [debouncedValue, setDebouncedValue] = useState(value);
  const [isPending, startTransition] = useTran();

  useEffect(() => {
    if (!updatable) return;

    const handler = setTimeout(() => {
      startTransition(() => setDebouncedValue(value));
    }, delay);
    return () => {
      clearTimeout(handler);
    };
  }, [value, delay, updatable]);

  return [debouncedValue, isPending];
}

// compatibility with older react versions
const useTran =
  useTransition ?? /* istanbul ignore next */ (() => [false, identityRun]);
const identityRun = (fun: (...args: any) => any) => fun();
