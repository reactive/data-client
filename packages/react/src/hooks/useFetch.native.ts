import { DenormalizeNullable, ExpiryStatus, NI } from '@data-client/core';
import {
  EndpointInterface,
  Denormalize,
  Schema,
  FetchFunction,
} from '@data-client/core';
import { useMemo } from 'react';
import { InteractionManager } from 'react-native';

import useCacheState from './useCacheState.js';
import useController from './useController.js';
import useFocusEffect from './useFocusEffect.native.js';

/**
 * Request a resource if it is not in cache.
 * @see https://dataclient.io/docs/api/useFetch
 */
export default function useFetch<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>]
): E['schema'] extends undefined | null ? ReturnType<E>
: Promise<Denormalize<E['schema']>>;

export default function useFetch<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>] | readonly [null]
): E['schema'] extends undefined | null ? ReturnType<E> | undefined
: Promise<DenormalizeNullable<E['schema']>>;

export default function useFetch<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>] | readonly [null]
): Promise<any> | undefined {
  const state = useCacheState();
  const controller = useController();

  const key = args[0] !== null ? endpoint.key(...args) : '';
  const cacheResults = key && state.endpoints[key];
  const meta = state.meta[key];

  // Compute denormalized value
  const { expiryStatus, expiresAt } = useMemo(() => {
    return controller.getResponseMeta(endpoint, ...args, state);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    cacheResults,
    state.indexes,
    state.entities,
    state.entitiesMeta,
    meta,
    key,
  ]);

  // If we are hard invalid we must fetch regardless of triggering or staleness
  const forceFetch = expiryStatus === ExpiryStatus.Invalid;

  const maybePromise = useMemo(() => {
    // null params mean don't do anything
    if ((Date.now() <= expiresAt && !forceFetch) || !key) return;
    // if args is [null], we won't get to this line
    return controller.fetch(endpoint, ...(args as Parameters<E>));
    // we need to check against serialized params, since params can change frequently
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [expiresAt, key, forceFetch, state.lastReset]);

  useFocusEffect(() => {
    // revalidating non-suspending data is low priority, so make sure it doesn't stutter animations
    const task = InteractionManager.runAfterInteractions(() => {
      if (Date.now() > expiresAt && key) {
        controller.fetch(endpoint, ...(args as Parameters<E>));
      }
    });

    return () => task.cancel();
  }, []);

  return maybePromise;
}
