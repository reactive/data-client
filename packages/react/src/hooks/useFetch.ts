import { DenormalizeNullable, ExpiryStatus, NI } from '@data-client/core';
import {
  EndpointInterface,
  Denormalize,
  Schema,
  FetchFunction,
} from '@data-client/core';
import { useMemo } from 'react';

import useCacheState from './useCacheState.js';
import useController from '../hooks/useController.js';

/**
 * Fetch an Endpoint if it is not in cache or stale.
 * @see https://dataclient.io/docs/api/useFetch
 */
export default function useFetch<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>]
): E['schema'] extends undefined | null ? ReturnType<E>
: Promise<Denormalize<E['schema']>>;

export default function useFetch<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>] | readonly [null]
): E['schema'] extends undefined | null ? ReturnType<E> | undefined
: Promise<DenormalizeNullable<E['schema']>>;

export default function useFetch<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>] | readonly [null]
): Promise<any> | undefined {
  const state = useCacheState();
  const controller = useController();

  const key = args[0] !== null ? endpoint.key(...args) : '';
  const cacheResults = key && state.endpoints[key];
  const meta = state.meta[key];

  // Compute denormalized value
  const { expiryStatus, expiresAt } = useMemo(() => {
    return controller.getResponseMeta(endpoint, ...args, state);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    cacheResults,
    state.indexes,
    state.entities,
    state.entitiesMeta,
    meta,
    key,
  ]);

  // If we are hard invalid we must fetch regardless of triggering or staleness
  const forceFetch = expiryStatus === ExpiryStatus.Invalid;

  return useMemo(() => {
    // null params mean don't do anything
    if ((Date.now() <= expiresAt && !forceFetch) || !key) return;

    // if args is [null], we won't get to this line
    return controller.fetch(endpoint, ...(args as Parameters<E>));
    // we need to check against serialized params, since params can change frequently
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [expiresAt, key, forceFetch, state.lastReset]);
}
