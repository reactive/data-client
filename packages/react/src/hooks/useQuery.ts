import type {
  DenormalizeNullable,
  NI,
  Queryable,
  SchemaArgs,
} from '@data-client/core';
import { useEffect, useMemo } from 'react';

import useCacheState from './useCacheState.js';
import useController from './useController.js';

/**
 * Query the store.
 *
 * `useQuery` results are globally memoized.
 * @see https://dataclient.io/docs/api/useQuery
 */
export default function useQuery<S extends Queryable>(
  schema: S,
  ...args: NI<SchemaArgs<S>>
  // INVALID results in undefined regardless of DenormalizeNullable
): DenormalizeNullable<S> | undefined {
  const state = useCacheState();
  const controller = useController();

  const key = JSON.stringify(args);

  // even though controller.get() is memoized, its memoization is more complex than
  // this so we layer it up to improve rerenders
  const { data, countRef } = useMemo(() => {
    return controller.getQueryMeta(schema, ...args, state);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [state.indexes, state.entities, key]);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(countRef, [data]);

  return data;
}
