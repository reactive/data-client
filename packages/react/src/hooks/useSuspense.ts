import { ExpiryStatus } from '@data-client/core';
import type {
  EndpointInterface,
  Denormalize,
  Schema,
  FetchFunction,
  DenormalizeNullable,
  ResolveType,
} from '@data-client/core';
import { useEffect, useMemo } from 'react';

import useCacheState from './useCacheState.js';
import useController from './useController.js';

/**
 * Ensure an endpoint is available.
 * Suspends until it is.
 *
 * `useSuspense` guarantees referential equality globally.
 * @see https://dataclient.io/docs/api/useSuspense
 * @throws {Promise} If data is not yet available.
 * @throws {NetworkError} If fetch fails.
 */
export default function useSuspense<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>]
): E['schema'] extends undefined | null ? ResolveType<E>
: Denormalize<E['schema']>;

export default function useSuspense<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(
  endpoint: E,
  ...args: readonly [...Parameters<E>] | readonly [null]
): E['schema'] extends undefined | null ? ResolveType<E> | undefined
: DenormalizeNullable<E['schema']>;

export default function useSuspense<
  E extends EndpointInterface<
    FetchFunction,
    Schema | undefined,
    undefined | false
  >,
>(endpoint: E, ...args: readonly [...Parameters<E>] | readonly [null]): any {
  const state = useCacheState();
  const controller = useController();

  const key = args[0] !== null ? endpoint.key(...args) : '';
  const cacheResults = key && state.endpoints[key];
  const meta = state.meta[key];

  // Compute denormalized value
  const { data, expiryStatus, expiresAt, countRef } = useMemo(() => {
    return controller.getResponseMeta(endpoint, ...args, state);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    cacheResults,
    state.indexes,
    state.entities,
    state.entitiesMeta,
    meta,
    key,
  ]);

  // If we are hard invalid we must fetch regardless of triggering or staleness
  const forceFetch = expiryStatus === ExpiryStatus.Invalid;

  const maybePromise = useMemo(() => {
    // null params mean don't do anything
    if ((Date.now() <= expiresAt && !forceFetch) || !key) return;

    return (
      controller
        // if args is [null], we won't get to this line
        .fetch(endpoint, ...(args as Parameters<E>))
        .catch(() => {})
    );
    // we need to check against serialized params, since params can change frequently
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [expiresAt, key, forceFetch, state.lastReset]);

  // fully "valid" data will not suspend even if it is not fresh
  if (expiryStatus !== ExpiryStatus.Valid && maybePromise) {
    throw maybePromise;
  }

  const error = controller.getError(endpoint, ...args, state);

  if (error) throw error;

  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(countRef, [data]);

  return data;
}
