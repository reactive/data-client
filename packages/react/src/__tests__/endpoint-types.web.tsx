import { CacheProvider } from '@data-client/react';
import { DataProvider as ExternalDataProvider } from '@data-client/react/redux';
import { TypedArticleResource } from '__tests__/new';
import nock from 'nock';

import { act, makeRenderDataClient } from '../../../test';
import { useController, useSuspense } from '../hooks';
import { payload, createPayload, users, nested } from '../test-fixtures';

function onError(e: any) {
  e.preventDefault();
}
beforeEach(() => {
  if (typeof addEventListener === 'function')
    addEventListener('error', onError);
});
afterEach(() => {
  if (typeof removeEventListener === 'function')
    removeEventListener('error', onError);
});

describe('endpoint types', () => {
  describe.each([
    ['CacheProvider', CacheProvider],
    ['ExternalDataProvider', ExternalDataProvider],
  ] as const)(`%s should enforce defined types`, (_, makeProvider) => {
    let renderDataClient: ReturnType<typeof makeRenderDataClient>;
    let mynock: nock.Scope;

    beforeEach(() => {
      nock(/.*/)
        .persist()
        .defaultReplyHeaders({
          'Access-Control-Allow-Origin': '*',
          'Content-Type': 'application/json',
        })
        .options(/.*/)
        .reply(200)
        .get(`/article-cooler/${payload.id}`)
        .reply(200, payload)
        .delete(`/article-cooler/${payload.id}`)
        .reply(204, '')
        .delete(`/article/${payload.id}`)
        .reply(200, {})
        .get(`/article-cooler/0`)
        .reply(403, {})
        .get(`/article-cooler/666`)
        .reply(200, '')
        .get(`/article-cooler/`)
        .reply(200, nested)
        .post(`/article-cooler/`)
        .reply(200, createPayload)
        .get(`/user/`)
        .reply(200, users)
        .get(/article-cooler\/.*/)
        .reply(404, 'not found')
        .put(`/article-cooler/${payload.id}`)
        .reply(200, (uri, body: any) => ({ ...payload, ...body }))
        .put(/article-cooler\/[^5].*/)
        .reply(404, 'not found');

      mynock = nock(/.*/).defaultReplyHeaders({
        'Access-Control-Allow-Origin': '*',
        'Content-Type': 'application/json',
      });
    });

    beforeEach(() => {
      renderDataClient = makeRenderDataClient(makeProvider);
    });
    afterEach(() => {
      nock.cleanAll();
    });

    let errorSpy: jest.Spied<typeof console.error>;
    afterEach(() => {
      errorSpy.mockRestore();
    });
    beforeEach(() => {
      errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {}); // Suppress console.error
    });

    it('should pass with exact params', async () => {
      const { result, waitForNextUpdate } = renderDataClient(() => {
        return useSuspense(TypedArticleResource.get, {
          id: payload.id,
        });
      });
      expect(result.current).toBeUndefined();
      await waitForNextUpdate();
      expect(result.current.title).toBe(payload.title);
      expect(errorSpy.mock.calls).toEqual([]);
    });

    it('should fail with improperly typed param', async () => {
      const { result, waitForNextUpdate } = renderDataClient(() => {
        // @ts-expect-error
        return useSuspense(TypedArticleResource.get, {
          id: 'abc ' as any as Date,
        });
      });
      expect(result.current).toBeUndefined();
      await waitForNextUpdate();
      expect(result.error).toBeDefined();
      expect((result.error as any).status).toBe(404);
    });

    it('should work with everything correct', async () => {
      const { result } = renderDataClient(() => {
        return useController().fetch;
      });
      await act(async () => {
        const a = await result.current(
          TypedArticleResource.update,
          { id: payload.id },
          { title: 'hi' },
        );
      });
      expect(errorSpy.mock.calls).toEqual([]);
    });

    it('types should strictly enforce with parameters that are any', async () => {
      const { result } = renderDataClient(() => {
        return useController().fetch;
      });
      () =>
        result.current(
          TypedArticleResource.anyparam,
          { id: payload.id },
          // @ts-expect-error
          { title: 'hi' },
        );
      () => result.current(TypedArticleResource.anyparam, { id: payload.id });
      expect(errorSpy.mock.calls).toEqual([]);
    });

    it('types should strictly enforce with body that are any', async () => {
      const { result, controller } = renderDataClient(() => {
        return useController().fetch;
      });
      () => TypedArticleResource.anybody({ id: payload.id }, { title: 'hi' });
      () =>
        controller.fetch(
          TypedArticleResource.anybody,
          { id: payload.id },
          {
            title: 'hi',
          },
        );

      () =>
        result.current(
          TypedArticleResource.anybody,
          { id: payload.id },
          {
            title: 'hi',
          },
        );
      // @ts-expect-error
      () => result.current(TypedArticleResource.anybody(), { id: payload.id });
      expect(errorSpy.mock.calls).toEqual([]);
    });

    it('should console.error on invalid payload', async () => {
      const { result } = renderDataClient(() => {
        return useController().fetch;
      });
      await result.current(
        TypedArticleResource.update,
        { id: payload.id },
        // @ts-expect-error
        { title2: 'hi' },
      );
      await result.current(
        TypedArticleResource.update,
        { id: payload.id },
        // @ts-expect-error
        { title: 5 },
      );
      expect(errorSpy.mock.calls).toMatchSnapshot();
    });

    it('should error on invalid params', async () => {
      const { result } = renderDataClient(() => {
        return useController().fetch;
      });

      let caught;
      await act(async () => {
        try {
          await result.current(
            TypedArticleResource.update,
            // @ts-expect-error
            'hi',
            { title: 'hi' },
          );
        } catch (error) {
          caught = error;
        }
      });

      expect(caught).toEqual(expect.any(Error));
    });
  });
});
