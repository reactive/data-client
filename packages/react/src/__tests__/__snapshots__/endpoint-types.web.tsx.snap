// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`endpoint types CacheProvider should enforce defined types should console.error on invalid payload 1`] = `[]`;

exports[`endpoint types ExternalDataProvider should enforce defined types should console.error on invalid payload 1`] = `[]`;
